sklearn/inspection/_permutation_importance.py

```diff
@@ -30,9 +40,31 @@ def _calculate_permutation_scores(
     # (memmap). X.copy() on the other hand is always guaranteed to return a
     # writable data-structure whose columns can be shuffled inplace.
     X_permuted = X.copy()
+    y_mod = y.copy()
+
+    # Draw from rows if not full data is used and subset X and y
+    if max_samples is not X.shape[0]:
+        row_indices = _generate_indices(
+            random_state=random_state,
+            bootstrap=False,
+            n_population=X.shape[0],
+            n_samples=max_samples,
+        )
+        if hasattr(X_permuted, "iloc"):
+            X_permuted = X_permuted.iloc[row_indices]
+            if hasattr(y_mod, "iloc"):
+                y_mod = y_mod.iloc[row_indices]
+            else:
+                y_mod = y_mod[row_indices]
+        else:
+            X_permuted = X_permuted[row_indices]
+            if hasattr(y_mod, "iloc"):
+                y_mod = y_mod.iloc[row_indices]
+            else:
+                y_mod = y_mod[row_indices]
```

Comment:
You should use the utils `_safe_indexing` that will make the indexing for you whatever the type `X`.